system/ee/ExpressionEngine/Service/LivePreview/LivePreview.php

```diff
@@ -137,20 +122,41 @@ public function preview($channel_id, $entry_id = null, $preview_url = null)
                 $uri = str_ireplace($parsed_url['host'], '', $uri);
             }
             $uri = trim($uri, '/');
-        } elseif ($entry->hasPageURI()) {
-            $uri = $entry->getPageURI();
-            ee()->uri->page_query_string = $entry->entry_id;
-            if (! $template_id) {
-                $template_id = $entry->getPageTemplateID();
-            }
-        } else {
-            // We want to avoid replacing `{url_title}` with an empty string since that
-            // can cause the wrong thing to render (like 404s).
-            if (empty($entry->url_title)) {
-                $entry->url_title = $entry->entry_id;
-            }
+        }
 
-            $uri = str_replace(['{url_title}', '{entry_id}'], [$entry->url_title, $entry->entry_id], $channel->preview_url);
+        if (empty($preview_url) || $prefer_system_preview === true) {
+            if (! empty($_POST['pages__pages_uri'])
+                && ! empty($_POST['pages__pages_template_id'])) {
+                //pages data passed with POST
+                $values = [
+                    'pages_uri' => $_POST['pages__pages_uri'],
+                    'pages_template_id' => $_POST['pages__pages_template_id'],
+                ];
+
+                $page_tab = new \Pages_tab();
+                $site_pages = $page_tab->prepareSitePagesData($entry, $values);
+
+                ee()->config->set_item('site_pages', $site_pages);
```

Comment:
I think this would be going a little contrary to what Live preview is.  I believe right now it's really just a preview of your changes before their saved... This is actually saving the URL for the page before it may be saved as a whole?  (Which could also result in a page going live in the navigation before it's been saved, as such it would 404) 
Thoughts? @intoeetive

Reply:
I think there's truth in what you're saying. But this code has always been there (I only moved it to different position). And I'm afraid there is currently no way to set the Page URL without actually setting it to config (the approach to Pages needs to be reworked really!)

Reply:
After reviewing this, and going back in git history to where it was added, I agree... This has actually been here for years... we will address when we revamp pages.